extensions/scala3/deployment/src/main/java/io/quarkus/scala3/deployment/Scala3CompilationProvider.java

```diff
@@ -0,0 +1,160 @@
+ *  reflection. This allows you to write code that will work against multiple
+ *  versions of dotty without recompilation.
+ *
+ *  The trade-off is that you can only pass a SimpleReporter to this method
+ *  and not a normal Reporter which is more powerful.
+ *
+ *  Usage example: [[https://github.com/lampepfl/dotty/tree/master/compiler/test/dotty/tools/dotc/InterfaceEntryPointTest.scala]]
+ *
+ *  @param args       Arguments to pass to the compiler.
+ *  @param simple     Used to log errors, warnings, and info messages.
+ *                    The default reporter is used if this is `null`.
+ *  @param callback   Used to execute custom code during the compilation
+ *                    process. No callbacks will be executed if this is `null`.
+ *  @return
+ */
+
+/**
+ * This just here so that tooling doesn't try to attach the above Javadoc to this method
+ */
+public class Scala3CompilationProvider implements CompilationProvider {
+    private final Logger log = Logger.getLogger(Scala3CompilationProvider.class);
+
+    @Override
+    public Set<String> handledExtensions() {
+        return Collections.singleton(".scala");
+    }
+
+    @Override
+    public void compile(Set<File> files, Context context) {
+        List<String> sources = files.stream()
+                .map(File::getAbsolutePath)
+                .collect(Collectors.toList());
+
+        String classpath = context.getClasspath().stream()
+                .map(File::getAbsolutePath)
+                .collect(Collectors.joining(File.pathSeparator));
+
+        List<String> dottyCompilerArgs = Arrays.asList(
+                "-d", context.getOutputDirectory().getAbsolutePath(),
+                "-classpath", classpath);
+
+        String[] sourcesWithCompilerArgs = Stream
+                .concat(sources.stream(), dottyCompilerArgs.stream())
+                .toArray(String[]::new);
+
+        CustomSimpleReporter reporter = new CustomSimpleReporter();
+        CustomCompilerCallback callback = new CustomCompilerCallback();
+
+        try {
+            // Reflect to get the Dotty compiler on the application's Classpath
+            Class<?> mainClass = Class.forName("dotty.tools.dotc.Main");
+            Method process = mainClass.getMethod("process", String[].class, SimpleReporter.class, CompilerCallback.class);
+            // Run the compiler by calling dotty.tools.dotc.Main.process
+            process.invoke(null, sourcesWithCompilerArgs, reporter, callback);
+        } catch (ClassNotFoundException e) {
+            // Class.forName()
+            e.printStackTrace();
+        } catch (NoSuchMethodException e) {
+            // mainClass.getMethod()
+            e.printStackTrace();
+        } catch (InvocationTargetException e) {
+            // process.invoke()
+            e.printStackTrace();
+        } catch (IllegalAccessException e) {
+            // process.invoke()
+            e.printStackTrace();
```

Comment:
I think we could use a multicatch here. There's little value to know which method throws the exception.

Not sure but should we use the logger instead of `printStackTrace`?

Reply:
I actually wasn't sure what to do in each of these scenarios -- not familiar with Java error-handling best practices.

Would it to be to just `log.error(e);`?